examples/two-chip.rs

```diff
@@ -0,0 +1,636 @@
+    /// the circuit.
+    advice: [Column<Advice>; 2],
+
+    // We need to create a permutation between our advice columns. This allows us to
+    // copy numbers within these columns from arbitrary rows, which we can use to load
+    // inputs into our instruction regions.
+    perm: Permutation,
+
+    // The selector for the public-input gate, which uses one of the advice columns.
+    s_pub: Selector,
+
+    add_config: AddConfig,
+    mul_config: MulConfig,
+}
+// ANCHOR END: field-config
+
+// ANCHOR: add-config
+#[derive(Clone, Debug)]
+struct AddConfig {
+    advice: [Column<Advice>; 2],
+    perm: Permutation,
+    s_add: Selector,
+}
+// ANCHOR_END: add-config
+
+// ANCHOR: mul-config
+#[derive(Clone, Debug)]
+struct MulConfig {
+    advice: [Column<Advice>; 2],
+    perm: Permutation,
+    s_mul: Selector,
+}
+// ANCHOR END: mul-config
+
+// ANCHOR: field-chip
+/// The top-level chip that will implement the `FieldInstructions`.
+struct FieldChip<F: FieldExt> {
+    config: FieldConfig,
+    _marker: PhantomData<F>,
+}
+// ANCHOR_END: field-chip
+
+// ANCHOR: add-chip
+struct AddChip<F: FieldExt> {
+    config: AddConfig,
+    _marker: PhantomData<F>,
+}
+// ANCHOR END: add-chip
+
+// ANCHOR: mul-chip
+struct MulChip<F: FieldExt> {
+    config: MulConfig,
+    _marker: PhantomData<F>,
+}
+// ANCHOR_END: mul-chip
+
+// ANCHOR: add-chip-trait-impl
+impl<F: FieldExt> Chip<F> for AddChip<F> {
+    type Config = AddConfig;
+    type Loaded = ();
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+// ANCHOR END: add-chip-trait-impl
+
+// ANCHOR: add-chip-impl
+impl<F: FieldExt> AddChip<F> {
+    fn construct(config: <Self as Chip<F>>::Config, _loaded: <Self as Chip<F>>::Loaded) -> Self {
+        Self {
+            config,
+            _marker: PhantomData,
+        }
+    }
+
+    fn configure(
+        meta: &mut ConstraintSystem<F>,
+        advice: [Column<Advice>; 2],
+        perm: Permutation,
+    ) -> <Self as Chip<F>>::Config {
+        let s_add = meta.selector();
+
+        // Define our addition gate!
+        meta.create_gate("add", |meta| {
+            let lhs = meta.query_advice(advice[0], Rotation::cur());
+            let rhs = meta.query_advice(advice[1], Rotation::cur());
+            let out = meta.query_advice(advice[0], Rotation::next());
+            let s_add = meta.query_selector(s_add, Rotation::cur());
+            s_add * (lhs + rhs + out * -F::one())
+        });
+
+        AddConfig {
+            advice,
+            perm,
+            s_add,
+        }
+    }
+}
+// ANCHOR END: add-chip-impl
+
+// ANCHOR: add-instructions-impl
+impl<F: FieldExt> AddInstructions<F> for FieldChip<F> {
+    type Num = Number<F>;
+    fn add(
+        &self,
+        layouter: impl Layouter<F>,
+        a: Self::Num,
+        b: Self::Num,
+    ) -> Result<Self::Num, Error> {
+        let config = self.config().add_config.clone();
```

Comment:
Is it possible to avoid these clones? I tried but the lifetime requirements defeated me.

Reply:
These ones, I think not (at least if we want the sub-cores to be usable on their own). We still have a large reduction in config cloning from this PR.